File: rxbinding/src/main/java/com/forceson/rxbinding/widget/RxAdapterView.java
package com.forceson.rxbinding.widget;

import android.widget.AdapterView;

import com.forceson.rxbinding.internal.Functions;

import java.util.concurrent.Callable;

import io.reactivex.Observable;
import io.reactivex.functions.Consumer;


/**
 * Created by son on 2020-01-17.
 */
public final class RxAdapterView {

    public static Observable<Integer> itemLongClicks(AdapterView<?> view) {
        return itemLongClicks(view, Functions.FUNC0_ALWAYS_TRUE);
    }

    private static Observable<Integer> itemLongClicks(AdapterView<?> view, Callable<Boolean> handled) {
        return new AdapterViewItemLongClickObservable(view, handled);
    }

    public static Observable<Integer> itemSelections(AdapterView<?> view) {
        return new AdapterViewItemSelectionObservable(view);
    }

    public static Observable<AdapterViewSelectionEvent> selectionEvents(AdapterView<?> view) {
        return new AdapterViewSelectionObservable(view);
    }

    public static Consumer<Integer> setSelection(final AdapterView<?> view) {
        return new Consumer<Integer>() {
            @Override
            public void accept(Integer value) {
                view.setSelection(value);
            }
        };
    }

    private RxAdapterView() {
        throw new AssertionError("No instances.");
    }
}

File: rxbinding-kotlin/src/main/java/com/forceson/rxbinding/kotlin/widget/RxAdapterView.kt
package com.forceson.rxbinding.kotlin.widget

import android.widget.Adapter
import android.widget.AdapterView
import com.forceson.rxbinding.widget.AdapterViewSelectionEvent
import com.forceson.rxbinding.widget.RxAdapterView
import io.reactivex.Observable
import io.reactivex.functions.Consumer

/**
 * Created by son on 2020-02-10.
 */

fun <T: Adapter> AdapterView<T>.itemSelections(): Observable<Int> = RxAdapterView.itemSelections(this)

fun <T: Adapter> AdapterView<T>.selectionEvents(): Observable<AdapterViewSelectionEvent> = RxAdapterView.selectionEvents(this)

fun <T: Adapter> AdapterView<T>.itemLongClicks(): Observable<Int> = RxAdapterView.itemLongClicks(this)

fun <T: Adapter> AdapterView<T>.selection(): Consumer<in Int> = RxAdapterView.setSelection(this)